packages/viewer/src/components/viewer-markup/viewer-markup.tsx

```diff
@@ -0,0 +1,385 @@
+  >;
+
+  /**
+   * Dispatched when a markup is removed, either through user
+   * interaction or programmatically.
+   */
+  @Event()
+  public markupRemoved!: EventEmitter<
+    HTMLVertexViewerMarkupArrowElement | HTMLVertexViewerMarkupCircleElement
+  >;
+
+  @Element()
+  private hostEl!: HTMLElement;
+
+  /**
+   * Adds a new markup as a child to this component. A new markup
+   * component will be created from the template specified by
+   * `arrow-template-id`, `circle-template-id`, or if undefined
+   * a default element will be created.
+   *
+   * @param markup The markup to add.
+   * @returns The markup element that was created.
+   * @see {@link ViewerMarkups.arrowTemplateId}
+   * @see {@link ViewerMarkups.circleTemplateId}
+   */
+  @Method()
+  public async addMarkup(
+    markup: Markup
+  ): Promise<
+    HTMLVertexViewerMarkupArrowElement | HTMLVertexViewerMarkupCircleElement
+  > {
+    if (markup instanceof ArrowMarkup) {
+      const { start, end, id } = markup;
+
+      const el = this.createArrowMarkupElement();
+      el.id = id;
+      el.start = start;
+      el.end = end;
+
+      this.updatePropsOnMarkup(el);
+      this.hostEl.appendChild(el);
+      this.markupAdded.emit(el);
+      return el;
+    } else if (markup instanceof CircleMarkup) {
+      const { bounds, id } = markup;
+
+      const el = this.createCircleMarkupElement();
+      el.id = id;
+      el.bounds = bounds;
+
+      this.updatePropsOnMarkup(el);
+      this.hostEl.appendChild(el);
+      return el;
+    } else {
+      throw new Error(`Cannot add markup. Unknown type '${markup}'.`);
+    }
+  }
+
+  /**
+   * Removes a markup with the given ID, and returns the HTML element
+   * associated to the markup. Returns `undefined` if no markup is
+   * found.
+   *
+   * @param id The ID of the markup to remove.
+   * @returns The markup element, or undefined.
+   */
+  @Method()
+  public async removeMarkup(
+    id: string
+  ): Promise<
+    | HTMLVertexViewerMarkupArrowElement
+    | HTMLVertexViewerMarkupCircleElement
+    | undefined
+  > {
+    const markups = await this.getMarkupElements();
+    const markup = markups.find((m) => m.id === id);
+
+    if (markup != null) {
+      markup.remove();
+      this.markupRemoved.emit(markup);
+    }
+
+    return markup;
+  }
+
+  /**
+   * Returns a list of markup elements that are children of this component.
+   *
+   * @returns A list of all markups.
+   * @see {@link ViewerMarkup.getMarkupElement}
+   */
+  @Method()
+  public async getMarkupElements(): Promise<
+    Array<
+      HTMLVertexViewerMarkupArrowElement | HTMLVertexViewerMarkupCircleElement
+    >
+  > {
+    const circleMarkup = Array.from(this.hostEl.children).filter(
+      isVertexViewerCircleMarkup
+    );
+    const arrowMarkup = Array.from(this.hostEl.children).filter(
+      isVertexViewerArrowMarkup
+    );
+
+    return [...arrowMarkup, ...circleMarkup];
```

Comment:
Can you use a `||` in your filter function. This way element order is preserved.

Reply:
For sure. I had originally run into some typing issues when using the `||` here, but that was probably because I had the return typing here incorrect. I'll update